bpi.module

```diff
@@ -857,3 +555,207 @@ function bpi_error_message($exception, $level) {
+  $delete_sid = $states_reversed['Deleted from BPI'];
+  $created_sid = $states_reversed['Created from BPI'];
+  $presend_sid = $states_reversed['Pre-sending to BPI'];
+
+  // Remove 'Save and push' button if the current state is already 'Sent to
+  // BPI'. We have to do this, because it's not possible to disallow workflow
+  // transitions to the current state.
+  if ($form['#default_value'] == $push_sid) {
+    unset($form['workflow']['workflow_sid']['#options'][$push_sid]);
+  }
+
+  // Remove "Delete from BPI" button, when we have already deleted from BPI.
+  if ($form['#default_value'] == $delete_sid) {
+    unset($form['workflow']['workflow_sid']['#options'][$delete_sid]);
+  }
+
+  // Remove "Save" button (saving to same workflow state) when the node is
+  // prepared for sending to BPI.
+  if ($form['#default_value'] == $presend_sid) {
+    unset($form['workflow']['workflow_sid']['#options'][$presend_sid]);
+  }
+
+  // We can't get the node id directly from the form, so we get it from the
+  // form_id which is generated with the node id.
+  $match = array();
+  preg_match('/workflow_transition_form_node_(\d+)/', $form['#form_id'], $match);
+  $nid = $match[1];
+  if ($nid > 0 && $form['#default_value'] != $created_sid) {
+    $form['bpi'] = bpi_http_push_action_form($nid);
+    array_unshift($form['#validate'], 'bpi_form_workflow_transition_form_validate');
+  }
+}
+
+/**
+ * Validate handler that will add custom submit handler.
+ *
+ * We add the submit handler here, because we need to add the submit handler,
+ * after the Workflow module has modified the form. Alternatively we could
+ * change the weight of the BPI module, but then we have to fix setting of
+ * $current_language in bpi_form_node_form_alter(), as changing the weight
+ * of the BPI module will make $current_language not be 'und', and BPI
+ * Syndication will fail. Also, we would have to test the module properly to
+ * be sure nothing else breaks.
+ */
+function bpi_form_workflow_transition_form_validate(&$form, &$form_state) {
+  $workflows = workflow_get_workflow_names();
+  $workflows_reversed = array_flip($workflows);
+  $wid = $workflows_reversed['BPI'];
+  $states = workflow_get_workflow_state_names($wid);
+  $states_reversed = array_flip($states);
+  $push_sid = $states_reversed['Sent to BPI'];
+  if (isset($form_state['triggering_element']['#workflow_sid']) && $form_state['triggering_element']['#workflow_sid'] == $push_sid) {
+    // The 'Save and push' button has been chosen, so we will check if the BPI
+    // values are correct.
+    if (empty($form_state['values']['bpi_push_category'])) {
+      form_set_error('bpi_push_category', t('Category must be set when pushing to BPI'));
+    }
+    if (empty($form_state['values']['bpi_push_audience'])) {
+      form_set_error('bpi_push_audience', t('Audience must be set when pushing to BPI'));
+    }
+  }
+  array_unshift($form_state['submit_handlers'], 'bpi_ding_news_node_form_submit');
```

Comment:
Adding this ding specific handler (bpi_ding_news_node_form_submit) prevents the module from being used by other sites (ereolen etc). This could be handle by using an hook and allow the ding_news module (and other modules) to implement a hook (ding_news_bpi_submit_handler).